File: src/07-challenge/solution.py
def fn(height: int, ornament: str, frequency: int) -> str:
    tree = ""
    asterisks_since_last_ornament = 0

    for i in range(height):
        line = " " * (height - i - 1)

        for i in range(len(line), height + i):
            if (asterisks_since_last_ornament + 1) % frequency:
                line += "*"
                asterisks_since_last_ornament += 1
                continue

            line += ornament
            asterisks_since_last_ornament = 0

        tree += f"{line}\n"

    trunk = f"{" " * max(height - 1, 0)}#"
    tree += trunk

    return tree


fn(5, "o", 2)

File: src/07-challenge/solution.ts
export default function fn(height: number, ornament: string, frequency: number): string {
	let tree: string = ''
	let asterisksSinceLastOrnament = 0

	for (let i = 0; i < height; i++) {
		let line = ' '.repeat(height - i - 1)

		for (let j = line.length; j < height + i; j++) {
			if ((asterisksSinceLastOrnament + 1) % frequency) {
				line += '*'
				asterisksSinceLastOrnament++
				continue
			}

			line += ornament
			asterisksSinceLastOrnament = 0
		}

		tree += `${line}\n`
	}

	const trunk = `${' '.repeat(Math.max(height - 1, 0))}#`
	tree += trunk

	return tree
}
